### ISSUE TITLE
visualising blend shapes and improving finer details

### ISSUE THREAD

**Issue by ketan-gan** [zjumsj/GaussianBlendshapes#13]
Hello, thank you for sharing your work! I wanted some suggestions on how to integrate this in my workflow.

- How to visualise the learned blend shapes, the paper mentions 50 blendshapes being created, what part of the code has these blend shape Gaussians, are these directly being used with flame blind weights to drive the avatar?
- What parameters to tinker with to improve finer details like facial hair? I would assume increasing `densify_until_iter` and decreasing `densify_grad_threshold` should help with that?
- for re-enactment, I found that it is best to find the closest flame blend weights from the training video and using those as the model does not generalise, but this requires some sort of smoothing as temporal stability is compromised, I tried using EMA on the vertex positions but that still gave me jittery results. Are there other parameters or perhaps something else that I should smoothen?

I would love to hear your thoughts on these, thanks!

**Comment by LinzhouLi** [comment]
1. There is no code to visualize each gaussian blendshape in these repo. But it is simple to implenment, you can just set blendshape weights to one hot encodings (such as `[1,0,0,0,...]`, `[0,1,0,0,...]`) and render the blendshape model. You can scale these weights and even set them to negative values to get the best visualization results.
2. If you want to improve the finer details you can try to decrease `densify_grad_threshold` to add more gaussians, or you can add `LPIPS` loss. If you found the contour of the reconstructed blendshape model is blur, you can try to remove the alpha loss and usign random background color during training.
3. Do you smooth the FLAME vertex positions during re-enactment? The blendshape model is driven by blendshape weights but not the FLAME mesh. Maybe smoothing the blendshape weights help with the jittery results.

**Comment by sw1014** [comment]
> 1. There is no code to visualize each gaussian blendshape in these repo. But it is simple to implenment, you can just set blendshape weights to one hot encodings (such as `[1,0,0,0,...]`, `[0,1,0,0,...]`) and render the blendshape model. You can scale these weights and even set them to negative values to get the best visualization results.
> 2. If you want to improve the finer details you can try to decrease `densify_grad_threshold` to add more gaussians, or you can add `LPIPS` loss. If you found the contour of the reconstructed blendshape model is blur, you can try to remove the alpha loss and usign random background color during training.
> 3. Do you smooth the FLAME vertex positions during re-enactment? The blendshape model is driven by blendshape weights but not the FLAME mesh. Maybe smoothing the blendshape weights help with the jittery results.

Hi, Shengjie, When I use one-hot embedding to represent blendshape weights and visualize each blendshapes, I find that the expression changes are very subtle, as shown in the image below. I would like to know the range of blend weights, as it seems that 1 is not sufficient to characterize the expression corresponding to each blendshapes.
I would be very grateful for any reply, thanks!!

![Image](https://github.com/user-attachments/assets/1dba7100-f6a1-4500-86fd-a1faa4386dfe)

**Comment by zjumsj** [comment]
You’re right — the scales of the FLAME blendshapes are inherently very small. When I did the visualization, I didn’t apply a fixed value across all expressions, but the values I used were generally larger, around 2 to 3.

I recommend using our visualization tool (https://github.com/zjumsj/FaceViewer). Go to **Settings → See Expr Basis** to reset all expressions to zero, and then use the sliders under **Expression** to edit the expressions you want.

**Comment by sw1014** [comment]
> You’re right — the scales of the FLAME blendshapes are inherently very small. When I did the visualization, I didn’t apply a fixed value across all expressions, but the values I used were generally larger, around 2 to 3.
> 
> I recommend using our visualization tool (https://github.com/zjumsj/FaceViewer). Go to **Settings → See Expr Basis** to reset all expressions to zero, and then use the sliders under **Expression** to edit the expressions you want.

Thank you for your response! I noticed that on the "bala" dataset, when adjusting the blendshape weights to between 2 and 3, significant errors in facial synthesis occur. Could this because the bala training set does not include more exaggerated expressions? Have you conducted experiments on datasets with more exaggerated expressions, and would this issue be better resolved in such cases?

**Comment by zjumsj** [comment]
A typical talking video alone is often not sufficient to cover a wide range of expressions. We've experimented with recording a 5-minute video that includes various predefined exaggerated expressions, and that significantly improves the model's generalization ability.

But unfortunately, due to FLAME using PCA decomposition and potentially lacking strong semantic meaning, some parameters may never take on large values within the training set. As a result, the corresponding Gaussian blendshapes may not be well optimized.  Expecting every blendshape to produce a visually reasonable result may not be realistic, but increasing the diversity of the training data can at least improve generalization performance during self-reenactment.

**Comment by sw1014** [comment]
> A typical talking video alone is often not sufficient to cover a wide range of expressions. We've experimented with recording a 5-minute video that includes various predefined exaggerated expressions, and that significantly improves the model's generalization ability.
> 
> But unfortunately, due to FLAME using PCA decomposition and potentially lacking strong semantic meaning, some parameters may never take on large values within the training set. As a result, the corresponding Gaussian blendshapes may not be well optimized. Expecting every blendshape to produce a visually reasonable result may not be practical, but increasing the diversity of the training data can at least improve generalization performance during self-reenactment.

Thank you for your explanation, it helped a lot!
